tests/test_utils.py

```diff
@@ -616,3 +617,23 @@ def test_num_samples(self, mock_stdout):
                 """),
         ]
         self.assertIn(output, possible_outputs)
+
+
+class TestRowiseEntropy(unittest.TestCase):
+    @parameterized.expand(
+        [
+            (dtype, chunk_size)
+            for dtype in (torch.float64, torch.float32, torch.float16, torch.bfloat16)
+            for chunk_size in (1, 16)
+        ]
+    )
+    def test_rowise_entropy(self, dtype, chunk_size):
+        batch_size, seq_len, vocab_size = 64, 384, 768
+        logits = torch.randn(batch_size, seq_len, vocab_size, dtype=dtype)
+        log_probs = torch.log_softmax(logits, dim=-1)
+        expected_entropy = -(torch.exp(log_probs) * log_probs).sum(-1)
+        predicted_entropy = rowise_entropy(logits, chunk_size=chunk_size)
+        if dtype in [torch.float16, torch.bfloat16]:
+            self.assertTrue(torch.equal(predicted_entropy, expected_entropy))
+        else:
+            torch.testing.assert_close(predicted_entropy, expected_entropy, rtol=1e-5, atol=1e-5)
```

Comment:
```suggestion
        if dtype in (torch.float64, torch.float32):
            p = logits.softmax(-1)
            entropy = -torch.sum(p * p.log(), dim=-1)
        else:
            logps = logits.log_softmax(dim=-1)
            entropy = -(torch.exp(logps) * logps).sum(-1)

        predicted_entropy = entropy_from_logits(logits, chunk_size=chunk_size)
        torch.testing.assert_close(predicted_entropy, entropy, rtol=1e-5, atol=1e-5)
```

we can use the exact calculation for float32/64